File: src/main/java/mods/cybercat/gigeresque/client/entity/texture/EggmorphLayerTexture.java
package mods.cybercat.gigeresque.client.entity.texture;

import com.mojang.blaze3d.platform.NativeImage;

import net.minecraft.client.Minecraft;
import net.minecraft.client.renderer.RenderType;
import net.minecraft.client.renderer.texture.DynamicTexture;
import net.minecraft.client.renderer.texture.TextureManager;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.packs.resources.ResourceManager;

public class EggmorphLayerTexture implements AutoCloseable {
	public RenderType renderLayer;
	private DynamicTexture texture;

	public EggmorphLayerTexture(TextureManager textureManager, ResourceManager resourceManager, ResourceLocation base) {
		texture = new DynamicTexture(new NativeImage(128, 128, true));
		Minecraft.getInstance().execute(new Runnable() {
			@Override
			public void run() {
				try {
					var baseImage = NativeImage.read(resourceManager.getResourceOrThrow(base).open());
					var layerImage = NativeImage
							.read(resourceManager.getResourceOrThrow(EntityTextures.EGGMORPH_OVERLAY).open());
					texture = new DynamicTexture(new NativeImage(layerImage.getWidth(), layerImage.getHeight(), true));
					var height = layerImage.getHeight();
					var width = layerImage.getWidth();
					var heightRatio = height / ((float) baseImage.getHeight());
					var widthRatio = width / ((float) baseImage.getWidth());
					for (int y = 0; y < height; y++) {
						for (int x = 0; x < width; x++) {
							var color = baseImage.getPixelRGBA(((int) (x / widthRatio)), ((int) (y / heightRatio)));
							var alpha = color >> 24 & 255;
							if (alpha > 25) {
								NativeImage image = texture.getPixels();
								if (image != null)
									image.setPixelRGBA(x, y, layerImage.getPixelRGBA(x, y));
							}
						}
					}
					texture.upload();
					baseImage.close();
				} catch (Exception e) {
				}
			}
		});
		ResourceLocation id = textureManager.register("eggmorph_layer/" + base.getPath(), texture);
		renderLayer = RenderType.entityTranslucent(id);
	}

	@Override
	public void close() throws Exception {
		texture.close();
	}
}

File: src/main/kotlin/com/bvanseg/gigeresque/client/entity/texture/EggmorphLayerTexture.kt
//package com.bvanseg.gigeresque.client.entity.texture
//
//import com.bvanseg.gigeresque.common.Gigeresque
//import net.fabricmc.api.EnvType
//import net.fabricmc.api.Environment
//import net.minecraft.client.MinecraftClient
//import net.minecraft.client.render.RenderLayer
//import net.minecraft.client.texture.NativeImage
//import net.minecraft.client.texture.NativeImageBackedTexture
//import net.minecraft.client.texture.TextureManager
//import net.minecraft.resource.ResourceManager
//import net.minecraft.util.Identifier
//
///**
// * @author Aelpecyem
// */
//@Environment(EnvType.CLIENT)
//class EggmorphLayerTexture(textureManager: TextureManager, resourceManager: ResourceManager, base: Identifier) :
//    AutoCloseable {
//
//    companion object {
//        private val layerTexture = Identifier(
//            Gigeresque.MOD_ID,
//            "textures/misc/eggmorph_overlay.png"
//        )
//    }
//
//    val renderLayer: RenderLayer
//    private var texture: NativeImageBackedTexture
//
//    override fun close() {
//        texture.close()
//    }
//
//    init {
//        texture = NativeImageBackedTexture(NativeImage(128, 128, true))
//        MinecraftClient.getInstance().execute {
//            try {
//                val baseImage = NativeImage.read(resourceManager.getResource(base).inputStream)
//                val layerImage = NativeImage.read(
//                    resourceManager.getResource(layerTexture).inputStream
//                )
//                texture = NativeImageBackedTexture(NativeImage(layerImage.width, layerImage.height, true))
//                val height = layerImage.height
//                val width = layerImage.width
//                val heightRatio = height / baseImage.height.toFloat()
//                val widthRatio = width / baseImage.width.toFloat()
//                for (y in 0 until height) {
//                    for (x in 0 until width) {
//                        val color = baseImage.getColor((x / widthRatio).toInt(), (y / heightRatio).toInt())
//                        val alpha = color shr 24 and 255
//                        if (alpha > 25) {
//                            texture.image?.setColor(x, y, layerImage.getColor(x, y))
//                        }
//                    }
//                }
//                texture.upload()
//                baseImage.close()
//            } catch (e: Exception) {
//                Gigeresque.LOGGER.error("Could not generate eggmorph texture for $base.", e)
//            }
//        }
//        val id = textureManager.registerDynamicTexture("eggmorph_layer/" + base.path, texture)
//        renderLayer = RenderLayer.getEntityTranslucent(id)
//    }
//}
